backpack/core/derivatives/nll_base.py

```diff
@@ -0,0 +1,132 @@
+from math import sqrt
+from typing import List, Tuple
+
+from torch import Size, Tensor, enable_grad, stack
+from torch.autograd import Variable, grad
+from torch.nn import CrossEntropyLoss
+
+from backpack.core.derivatives.basederivatives import BaseLossDerivatives
+from backpack.utils.subsampling import subsample
+
+
+class NLLLossDerivatives(BaseLossDerivatives):
+    """Partial derivative bases for NLL loss."""
+
+    def _sqrt_hessian_sampled(
+        self,
+        module: CrossEntropyLoss,
+        g_inp: Tuple[Tensor],
+        g_out: Tuple[Tensor],
+        mc_samples: int = 1,
+        subsampling: List[int] = None,
+    ) -> Tensor:
+        """Method to approximate the square root Hessian through Monte Carlo sampling.
+
+        For a given loss, either _make_distribution or compute_sampled_grads must be
+        implemented.
+
+        Optionally, it is possible to specify _checks to perform before calculating
+        the loss gradient.
+
+        For use in mean mode, _get_mean_normalization must be implemented.
+
+        Args:
+            module:
+            g_inp:
+            g_out:
+            mc_samples: number of Monte Carlo samples to take
+            subsampling: Indices of samples that are sliced along the dimension.
+            Defaults to ``None`` (use all samples).
+
+        Returns:
+            Monte Carlo sampled gradients
+
+        """
+        self._checks(module)
+        subsampled_input = Variable(
+            subsample(module.input0, subsampling=subsampling), requires_grad=True
+        )
+        sampled_grads = self.compute_sampled_grads(subsampled_input, mc_samples) / sqrt(
+            mc_samples
+        )
+        if module.reduction == "mean":
+            sampled_grads /= sqrt(self._get_mean_normalization(module.input0))
+        return sampled_grads
+
+    def _checks(self, module):
+        """Any checks to be performed. Default runs none."""
+        return
+
+    def compute_sampled_grads(self, subsampled_input, mc_samples):
+        """Method to create the sampled gradients.
+
+        For loss functions that can be expressed as a Negative Log Likelihood (NLL), the
+        Maximum a-Posteriori (MAP) estimate of the network parameters can be written as
+        𝜃ₘₐₚ = argmin_𝜃 (𝑟(𝜃)+𝑙(xₙ,yₙ;𝜃)) where 𝑟 is the regularizer and 𝑙 is the loss
+        function, defined here as 𝑙(xₙ,yₙ;𝜃)= −log p(yₙ | f_𝜃(xₙ)). TODO how to cite this?
+
+        compute_sampled_grads calculates and returns the gradient of the loss with respect
+        to each of the randomly drawn samples. To use this function, the user must implement
+        the functions _make_distribution.
+
+        Alternatively, if a faster gradient calculation is known for the loss function, the
+        user may overwrite this method entirely with a hand-crafted method. Then the method
+        should return the gradient of the loss with respect to the subsampled input for each
+        of the Monte Carlo samples
```

Comment:
Suggestion: By default, this will compute gradients for samples of the the likelihood distribution with  `autograd`. This function can be overwritten if the gradient is known analytically.

Reply:
Fixed this in commit
@jabader97 [Implemented (some of) commented fixes](https://github.com/f-dangel/backpack/pull/252/commits/8201fc6f73bf7231f82779462f3dc46034d21c5e)